### PR TITLE
BAAS-20006: Update goja to latest (March 2023) - Part 2

### PR DESCRIPTION
**Pull request by Gabri3l**
### [Implemented classes (including private elements).](https://github.com/mongodb-forks/goja/pull/80/commits/aeab1c09fc0f96e5506d5034f9ff333dc35bbc44)

* Added `call.ctx` to `builtin_function.go` on line 198. It won't look like a change because I ported it over from before.
* Updated signature of `f._call` to now take a context param too. This was necessary because we don't pass `call` as an argument anymore but we pass the function call arguments directly.
* Updated `common_eval` signature base on the new `eval` function signature.
* Generally moved `MemUsage` functions around, nothing particularly crazy.

### [Fixed "delete 'a'.prop". Optimised a[<const_string>] to a.<const_string>](https://github.com/mongodb-forks/goja/pull/80/commits/b5ce5dc0c9a26454ac6684b459bc896851b4ec85)

The “array with non-numeric keys" needed a change. There was an optimization introduced with dot notation, before we had two string props:

* `x["a"] = 3`
* `x["c"] = 3`

Both were the last item added on the stack before we check the memory the only difference between the two mem checks was the added "abc" and the added number.

After the change this is not true anymore and elements that can use dot notation are not added to the stack anymore. This simply means that "abc" is now the last item on the stack before we call the mem check again which would have resulted in additional mem increase of 3 (i.e. `len("abc")`). To "even things out" the last item added before the mem check are assigned with a number key value which will isolate again "abc" and the added number so the resulting mem increase stays the same.

### [Fixed Export() value for arrow functions and classes. Added Runtime.A…](https://github.com/mongodb-forks/goja/pull/80/commits/cb7a47c68859e671d8569feda1a52f5e66b82dbd)

* There was some refactoring due to some shared code between `AssertFunction` and `AssertConstructor` which was encapsulated in `runWrapped`. I just had to make sure we were still passing the correct context along.

### [Refactored conversion to primitive. Always use String() method if d…](https://github.com/mongodb-forks/goja/pull/80/commits/3fbf936642aa5a21f3d54b046e2616778fcb6aa9)

* Had to remove a few unused methods in object_goreflect and object_goslice

## BAAS

We have an evergreen patch [here](https://spruce.mongodb.com/version/641472f6d1fe079d32bba27c/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) that's looking promising. The failing tests seem unrelated.

### CODE REVIEW

**Comment by Gabri3l** [comment]
`evergreen refresh`

**Comment by Gabri3l** [comment]
evergreen restart

File: builtin_map.go [hunk 1]
```diff
@@ -68,9 +68,7 @@ func (mo *mapObject) export(ctx *objectExportCtx) interface{} {
 }
 
 func (mo *mapObject) exportToMap(dst reflect.Value, typ reflect.Type, ctx *objectExportCtx) error {
-	if dst.IsNil() {
-		dst.Set(reflect.MakeMap(typ))
-	}
+	dst.Set(reflect.MakeMap(typ))
```
Comment (kpatel71716) [inline builtin_map.go:71]: [q] curious why this nil check was removed. Is that not something we consider possible?

File: builtin_set.go [hunk 2]
```diff
@@ -89,6 +89,7 @@ func (so *setObject) exportToArrayOrSlice(dst reflect.Value, typ reflect.Type, c
 }
 
 func (so *setObject) exportToMap(dst reflect.Value, typ reflect.Type, ctx *objectExportCtx) error {
+	dst.Set(reflect.MakeMap(typ))
```
Comment (kpatel71716) [inline builtin_set.go:92]: [q] similar q as above. We don't expect this to be nil, correct?

File: vm.go [hunk 3]
```diff
@@ -1822,26 +2221,60 @@ func (g getProp) exec(vm *vm) {
...
+
+type getPropCallee unistring.String
+
+func (g getPropCallee) exec(vm *vm) {
+	v := vm.stack[vm.sp-1]
+	obj := v.baseObject(vm.r)
+	n := unistring.String(g)
+	if obj == nil {
+		// (REALMC-7469) can revert this once otto is gone
+		panic(vm.r.NewTypeError("Cannot access member '%s' of undefined", n))
```
Comment (kpatel71716) [inline vm.go:2266]: [q] I thought we can revert this now? Fine to do it in a separate PR though

**Review by kpatel71716 — APPROVED** [review]
LGTM! Some qs but no blockers

**Review by LijieZhang1998 — APPROVED** [review]
I'll blindly approve it cause I don't want to block it.

File: builtin_map.go (hunk 1 above)
Comment (Gabri3l) [reply builtin_map.go:71]: I think the gist of it is that it doesn't matter anymore and they want to create a map when exporting no matter what (nil or not nil), they cover this a bit in the new test. Before it would only create it if the destination map was nil, now it doesn't matter. It will overwrite, if you look at `TestMapExportToNonNilMap` before this change that test would fail as the map `m` would not be overwritten, this this change it does. So if you were to add a check like:

```go
func TestMapExportToNonNilMap(t *testing.T) {
	vm := New()
	m := map[int]interface{}{
		2: true,
	}
	res, err := vm.RunString("new Map([[1, true]])")
	if err != nil {
		t.Fatal(err)
	}
	err = vm.ExportTo(res, &m)
	if err != nil {
		t.Fatal(err)
	}
	if len(m) != 1 {
		t.Fatal(m)
	}
        // THIS IS DIFFERENT TO UNDERSTAND
	// if _, exists := m[1]; !exists {
	//	 t.Fatal(m)
	// }
	if _, exists := m[2]; !exists {
		t.Fatal(m)
	}
}
```

It now fails but it would have passed before this change.

File: compiler.go [hunk 4]
```diff
@@ -576,56 +729,84 @@ func (s *scope) finaliseVarAlloc(stackOffset int) (stashSize, stackSize int) {
...
 				base := scope.base
-				if argsInStash {
+				if this {
+					if derivedCtor {
+						for _, pc := range *aps {
+							ap := &code[base+pc]
+							switch (*ap).(type) {
+							case loadStack:
+								*ap = loadThisStack{}
+							case initStack:
```
Comment (jwongmongodb) [inline compiler.go:742]: [opt] we can group the no-ops together in the switch statement

File: compiler.go [hunk 5]
```diff
@@ -576,56 +729,84 @@ func (s *scope) finaliseVarAlloc(stackOffset int) (stashSize, stackSize int) {
 				}
 				code := scope.prg.code
 				base := scope.base
-				if argsInStash {
+				if this {
+					if derivedCtor {
```
Comment (jwongmongodb) [inline compiler.go:736]: [opt] can combined this with the above check 

```
if this && derivedCtor{
...
}
```

**Review by jwongmongodb — APPROVED** [review]
Thanks for breaking up to the commits. I don't have much to add for the PR, just a few optional cosmetic fixes since it came from the original goja code

File: compiler.go (hunk 5 above)
Comment (Gabri3l) [reply compiler.go:736]: Both very legitimate changes! I'm only NOT going to do it to minimize conflicts in the future. Although our diversions are not clear yet (working on it ...) those are the only places where we can 100% make these type of changes. Much appreciate either way!
